File: terrain/textures/perlin.js
class Vector {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    // Returns the sum of x and y (matches your Python return_vector, but see note below)
    return_vector() {
        return this.x + this.y;
    }
}

function shuffle(arrayShuffle) {
    for (let x = 0; x < arrayShuffle.length - 1; x++) {
        let index = Math.floor(Math.random() * (x + 1));
        let temp = arrayShuffle[x];
        arrayShuffle[x] = arrayShuffle[index];
        arrayShuffle[index] = temp;
    }
}

function MakePermutation() {
    let permutation = [];
    for (let i = 0; i < 256; i++) {
        permutation.push(i);
    }

    shuffle(permutation);

    for (let i = 0; i < 256; i++) {
        permutation.push(permutation[i]);
    }

    return permutation;
}

const Permutation = MakePermutation();

function getVector(v) {
    let h = v & 3;

    if (h === 0) {
        return new Vector(1, 1);
    } else if (h === 1) {
        return new Vector(-1, 1);
    } else if (h === 2) {
        return new Vector(-1, -1);
    } else {
        return new Vector(1, -1);
    }
}

function fade(t) {
    return ((6 * t - 15) * t + 10) * Math.pow(t, 3);
}

function lerp(t, a, b) {
    return a + t * (b - a);
}

function Noise2D(x, y) {
    let X = Math.floor(x) & 255;
    let Y = Math.floor(y) & 255;
    let xf = x - Math.floor(x);
    let yf = y - Math.floor(y);

    let topRight = new Vector(xf - 1, yf - 1);
    let topLeft = new Vector(xf, yf - 1);
    let bottomRight = new Vector(xf - 1, yf);
    let bottomLeft = new Vector(xf, yf);

    let valueTopRight = Permutation[Permutation[X + 1] + Y + 1];
    let valueTopLeft = Permutation[Permutation[X] + Y + 1];
    let valueBottomRight = Permutation[Permutation[X + 1] + Y];
    let valueBottomLeft = Permutation[Permutation[X] + Y];

    // Dot product between gradient and distance vector
    function dot(a, b) {
        return a.x * b.x + a.y * b.y;
    }

    let dotTopRight = dot(getVector(valueTopRight), topRight);
    let dotTopLeft = dot(getVector(valueTopLeft), topLeft);
    let dotBottomRight = dot(getVector(valueBottomRight), bottomRight);
    let dotBottomLeft = dot(getVector(valueBottomLeft), bottomLeft);

    let u = fade(xf);
    let v = fade(yf);

    return lerp(u, lerp(v, dotBottomLeft, dotTopLeft), lerp(v, dotBottomRight, dotTopRight));
}

function brownianfractal(x, y, numOctaves) {
    let result = 0;
    let amplitude = 1;
    let frequency = 0.005;

    for (let i = 0; i < numOctaves; i++) {
        result += Noise2D(x * frequency, y * frequency) * amplitude;
        amplitude *= 0.5;
        frequency *= 2;
    }

    return result;
}

function perlinRender(xrange, yrange, o) {
    let colorMatrix = [];
    for (let y = 0; y < yrange; y++) {
        colorMatrix[y] = [];
        for (let x = 0; x < xrange; x++) {
            let n = brownianfractal(x, y, o);
            n = (n + 1) / 2;

            let c = Math.floor(n * 255);
            colorMatrix[y][x] = { r: c, g: c, b: c };
        }
    }
    return colorMatrix;
} 

export { perlinRender, /* any other functions */ };

File: terrain/textures/perlin.py
import random

class Vector:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def return_vector(self):
        return self.x + self.y
    
def shuffle(arrayShuffle):
    for x in range (0, len(arrayShuffle) - 1):
        index = int(random.random() * (x+1))
        temp = arrayShuffle[x]

        arrayShuffle[x] = arrayShuffle[index]
        arrayShuffle[index] = temp

def MakePermutation():
    permutation = []
    for i in range(256):
        permutation.append(i)
    
    shuffle(permutation)

    for i in range(256):
        permutation.append(permutation[i])

    return permutation

Permutation = MakePermutation()

def getVector(v):
    h = v & 3

    if h == 0:
        return Vector(1,1)
    elif h == 1:
        return Vector(-1,1)
    elif h == 2:
        return Vector(-1,-1)
    else:
        return Vector(1,-1)
    
def fade(t):
    return ((6*t-15)*t+10) * pow (t,3)

def lerp(t, a, b):
    return a+t*(b-a)

def Noise2D(x,y):
    X = int(x) & 255
    Y = int(y) & 255
    xf = x - int(x)
    yf = y - int(y)

    topRight = Vector(xf-1, yf-1)
    topLeft = Vector(xf, yf-1)
    bottomRight = Vector(xf-1, yf)
    bottomLeft = Vector(xf, yf)

    valueTopRight = Permutation[Permutation[X+1]+Y+1]
    valueTopLeft = Permutation[Permutation[X]+Y+1]
    valueBottomRight = Permutation[Permutation[X+1]+Y]
    valueBottomLeft = Permutation[Permutation[X]+Y]

    dotTopRight = topRight.return_vector(getVector(valueTopRight))
    dotTopLeft = topLeft.return_vector(getVector(valueTopLeft))
    dotBottomRight = bottomRight.return_vector(getVector(valueBottomRight))
    dotBottomLeft = bottomLeft.return_vector(getVector(valueBottomLeft))

    u = fade(xf)
    v = fade(yf)

    return lerp(u, lerp(v, dotBottomLeft, dotTopLeft), lerp(v, dotBottomRight, dotTopRight))

def brownianfractal(x,y,numOctaves):
    result = 0
    amplitude = 1
    frequency = 0.005

    for i in range(numOctaves):
        result += Noise2D(x * frequency, y * frequency) * amplitude
        amplitude *= 0.5
        frequency *= 2

    return result

def perlinRender(xrange, yrange, o):
    colorMatrix = [[0 for x in range(0, xrange)] for y in range(0, yrange)]
    for y in range(0, yrange):
        for x in range(0, xrange):
            n = brownianfractal(x,y,o)
            n = (n + 1) / 2

            c = int(n * 255)
            colorMatrix[y][x] = {'r': c, 'g': c, 'b': c}
    return colorMatrix

